### ISSUE TITLE
Embark's standard lints

### ISSUE THREAD

**Issue by repi** [EmbarkStudios/rust-ecosystem#59]
This issue tracks which Rust and Clippy lints we currently use as default for all of our internal and external crates. We have a common base of lints that we configure the same everywhere that is more opinionated than the Rust and Clippy defaults and geared towards how we prefer to (currently) write & maintain Rust code. 

Individual crates are able to, and often do override some of these defaults, either on a crate level, or for specific modules or functions. But we still find it helpful to have a standard across everything to keep the code high quality and generally working the same way.

We hope to later in 2021 have support for setting this configuration a shared clippy toml file across all of our repositories (tracked in #22, and https://github.com/rust-lang/cargo/issues/5034), but in the meantime we have this versioned code snippet that we use in our crate root files (`lib.rs` / `main.rs`):

## Current lint config

```
// BEGIN - Embark standard lints v0.2.
// do not change or add/remove here, but one can add exceptions after this section
#![deny(unsafe_code)]
#![warn(
    clippy::all,
    clippy::await_holding_lock,
    clippy::dbg_macro,
    clippy::debug_assert_with_mut_call,
    clippy::doc_markdown,
    clippy::empty_enum,
    clippy::enum_glob_use,
    clippy::exit,
    clippy::explicit_into_iter_loop,
    clippy::filter_map_next,
    clippy::fn_params_excessive_bools,
    clippy::if_let_mutex,
    clippy::imprecise_flops,
    clippy::inefficient_to_string,
    clippy::let_unit_value,
    clippy::linkedlist,
    clippy::lossy_float_literal,
    clippy::macro_use_imports,
    clippy::map_flatten,
    clippy::map_unwrap_or,
    clippy::match_on_vec_items,
    clippy::match_wildcard_for_single_variants,
    clippy::mem_forget,
    clippy::mismatched_target_os,
    clippy::needless_borrow,
    clippy::needless_continue,
    clippy::option_option,
    clippy::pub_enum_variant_names,
    clippy::ref_option_ref,
    clippy::rest_pat_in_fully_bound_structs,
    clippy::string_to_string,
    clippy::suboptimal_flops,
    clippy::todo,
    clippy::unnested_or_patterns,
    clippy::unused_self,
    clippy::verbose_file_reads,
    future_incompatible,
    nonstandard_style,
    rust_2018_idioms
)]
// END - Embark standard lints v0.2
// crate-specific exceptions:
#![allow()]
```

**Comment by repi** [comment]
Filed https://github.com/EmbarkStudios/opensource/issues/35 to start using the template in all of our current open source repos

**Comment by repi** [comment]
#60 is a PR draft and discussion about potential additions and removals for v0.3

**Comment by aron-granberg** [comment]
Personally I really don't like suboptimal_flops. In particular because it forces you to use `a.mul_add(2.0, 4.0)` instead of `a * 2.0 + 4.0` which imo is a lot more readable than the former.

**Comment by repi** [comment]
We've released v0.4 with significant amount of smaller pedantic lints added, and a couple removals. Details in PR #66

**Comment by Jake-Shadle** [comment]
I think we should remove `disallowed_type(s)` and `disallowed_method(s)` from the lint list, these lints only apply if they are actually configured for the repo, which a majority don't do, and it's also complicated because they were renamed in 1.59.0 to be plural, which means updating the list could make running clippy on older versions fail and require an update to all users of the lints since the renamed lints are now a warning.

**Comment by MarijnS95** [comment]
@Jake-Shadle There was a similar Rust 1.55+ requirement for v5, seems like we can do the same for a v6 release?

In any case, I fixed these in https://github.com/EmbarkStudios/rust-ecosystem/pull/73, we at traverse are actively using this setup. Looking at it again I'll remove them in that PR, as they're already warn-by-default and thus provide no extra benefit being configured explicit (if they weren't, `clippy.toml` config would be a useless no-op...).

**Comment by Jake-Shadle** [comment]
Yes, since they are warn by default I don't think it makes sense to keep them in the list at all, technically it's a breaking change to _keep_ them in the list due to the clippy change, but it will be fine to do that in a 0.6 version, I just wanted to post here for when that does happen because I'm in the process of removing them from all of the crates I maintain.

**Comment by repi** [comment]
Oh they are warn-by-default now, that's great and yes then we shall simply remove them. Will do a v6 release now soon requires 1.59 with them removed. And follow up with another release soon with new additional lints that we've been using and testing.

**Comment by repi** [comment]
Released v6 now with the renamed warn-by-default lints removed. (#74)

**Comment by MarijnS95** [comment]
Well, guess I am too late to update that PR :grimacing:

Any other lints in the list that are now default-warn?

With additional lints pending, shall we submit some of ours again too (after deduplicating with what Embark has been testing with)? We've been finding some interesting new flags as well.

**Comment by repi** [comment]
@MarijnS95 just wanted to get out v6 quickly to unblock Rust 1.59 clippy compile errors.

created #75 to discuss v7 additions and removals and good idea also to go through and see if there are more warn-by-default lints we can remove. and great if you have any other lints that you've been testing with and that works well, so let's discuss in that!

**Comment by epage** [comment]
I've created an RFC for package-level lint configuration: https://github.com/rust-lang/rfcs/pull/3389

**Comment by epage** [comment]
I feel like the above RFC has shaped up quite well.  As Embark has a large lint configuration that gets copied from project to project, I'm curious how well you feel the current design works out from the "large user" perspective.

In particular
- The `priority` field for controlling lint order (so you can enable a lint group like `all` and then override lints within it)
- This being focused more on linters and not targeting rustfmt and other tools
- The format being `<lint> = "<level>"` in the short form and `<lint> = { level = "<level>", priority = 0 }` in the long form
- Using workspace inheritance for explicitly sharing across the workspace
- Not allowing overriding of the workspace-inherited config in the initial release
- Not supporting lint configuration in the initial release

(I think that covered all the more controversial items)

**Comment by epage** [comment]
@repi there is one particular point the cargo team is hoping Embark can spare some time to provide feedback on regarding your experience with your lints lists.

Problem: RUSTFLAGS has baked in an ordering so users can have lints override groups.    We are exploring how to support this within the `[lints]` table.  A particularly challenging point is that there are no guarantees today around lint groups, e.g. two rustc lint groups intersect which makes it so we can't perfectly order lint groups on the users behalf.

We are particularly concerned about the experience managing large lists of lints, not just that some experts can maintain it but with an eye of empathy towards what problems other users may have.

In [a recent post](https://github.com/rust-lang/rfcs/pull/3389#discussion_r1157777423), I've laid out the options we are considering.   "large lists of lints" came up in two parts of the conversation within the cargo team
- One option more directly imitates RUSTFLAGS (Option 4, particularly variant 3), so you have direct experience with this. What are trying to get a feel for is wow likely are people to not explicitly track the groups within the list and sort across lints and groups, rather than just within lints and groups only if its safe
- If we make the format `warn  = [ "lint" ]`, we are wondering how likely might a user lose track of which list they are in (we've seen this with dependency tables)
  - The RUSTFLAGS one is what motivated for me to reach out, I'm mostly including this for completeness of discussing large lint list problems
